### PR TITLE
Add patterns to handle vector distribution for contract op with loop

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Support distributing op within a reduction loop

### CODE REVIEW

File: lib/Transforms/VectorDistribution.cpp [hunk 1]
```diff
@@ -20,6 +20,205 @@ using namespace mlir;
 using namespace mlir::linalg;
 
 namespace {
+
+/// Modified version of the upstream pattern handling more cases.
+/// TODO: Replace upstream once it is ready.
+struct TransferReadExtractPattern
```
Comment (nicolasvasilache) [inline lib/Transforms/VectorDistribution.cpp:26]: some docs please

Comment (nicolasvasilache) [reply lib/Transforms/VectorDistribution.cpp:26]: (here and below)

File: lib/Transforms/VectorDistribution.cpp [hunk 2]
```diff
@@ -20,6 +20,205 @@ using namespace mlir;
...
+    auto read = extract.vector().getDefiningOp<vector::TransferReadOp>();
+    if (!read || read.mask()) return failure();
+
+    SmallVector<Value, 4> indices(read.indices().begin(), read.indices().end());
+    AffineMap indexMap = extract.map().compose(read.permutation_map());
+    unsigned idCount = 0;
+    // We cannot pull down transfer_read ops as it may cross ops with
+    // side-effects.
+    rewriter.setInsertionPoint(read);
+    ImplicitLocOpBuilder lb(read.getLoc(), rewriter);
```
Comment (nicolasvasilache) [inline lib/Transforms/VectorDistribution.cpp:41]: no builder within a rewriter please, messes up the tracking (and our future drivers) IIRC

File: lib/Transforms/VectorDistribution.cpp [hunk 3]
```diff
@@ -20,6 +20,205 @@ using namespace mlir;
...
+    for (auto it :
+         llvm::zip(indexMap.getResults(), extract.map().getResults())) {
+      AffineExpr d0, d1;
+      bindDims(read.getContext(), d0, d1);
+      auto indexExpr = std::get<0>(it).dyn_cast<AffineDimExpr>();
+      if (!indexExpr) continue;
```
Comment (nicolasvasilache) [inline lib/Transforms/VectorDistribution.cpp:47]: comments and spacing please, it's a bit hard to review atm

File: lib/Transforms/VectorDistribution.cpp [hunk 4]
```diff
@@ -20,6 +20,205 @@ using namespace mlir;
...
+    yieldOp.getOperation()->setOperands(returnValues);
+    rewriter.replaceOp(
+        forOp, ArrayRef<Value>(repResults).take_front(forOp.getNumResults()));
+    return success();
+  }
+};
+
+/// Pattern to remove dead code in scf.for op after incremental conversion of
+/// for result values.
+/// Upstreaming this pass is under progress and can be replaced once it landes:
```
Comment (nicolasvasilache) [inline lib/Transforms/VectorDistribution.cpp:130]: nice thanks for being proactive here!

**Review by nicolasvasilache — CHANGES_REQUESTED** [review]
needs more docs to be reviewable, sorry

File: lib/Transforms/VectorDistribution.cpp (hunk 1 above)
Comment (ThomasRaoux) [reply lib/Transforms/VectorDistribution.cpp:40]: added docs for the different patterns plus extra comments.

File: lib/Transforms/VectorDistribution.cpp (hunk 3 above)
Comment (ThomasRaoux) [reply lib/Transforms/VectorDistribution.cpp:63]: I added extra comments, hopefully it is easier to read

**Review by ThomasRaoux — COMMENTED** [review]
Added comments to make it easier to understand.

File: lib/Transforms/VectorDistribution.cpp (hunk 2 above)
Comment (ThomasRaoux) [reply lib/Transforms/VectorDistribution.cpp]: removed

**Pull request closed without merging**